lib/puppet/provider/package/pip.rb

```diff
@@ -127,7 +127,11 @@ def latest
     if self.class.compare_pip_versions(command_version, '1.5.4') == -1
       available_versions_with_old_pip.last
     else
-      available_versions_with_new_pip.last
+      if self.class.compare_pip_versions(command_version, '20.2.4') == 1 &&
+        self.class.compare_pip_versions(command_version, '21.1') == -1
+        extra_arg = '--use-deprecated=legacy-resolver'
+      end
```

Comment:
To avoid duplication, you could extract this in a private method like:
```ruby

  private

  def list_extra_flags(command_version)
    klass = self.class
    if klass.compare_pip_versions(command_version, '20.2.4') == 1 &&
      klass.compare_pip_versions(command_version, '21.1') == -1
      '--use-deprecated=legacy-resolver'
    end
  end
```

and pass it to `available_versions_with_new_pip` like `available_versions_with_new_pip(list_extra_flags(command_version))`

To avoid duplication